src/Microsoft.ML.Transforms/UngroupTransform.cs

```diff
@@ -254,22 +259,47 @@ public SchemaImpl(IExceptionContext ectx, Schema inputSchema, UngroupMode mode,
                 _inputSchema = inputSchema;
                 Mode = mode;
 
-                CheckAndBind(_ectx, inputSchema, pivotColumns, out _infos);
+                Bind(_ectx, inputSchema, pivotColumns, out _infos);
 
-                _pivotColMap = new Dictionary<string, int>();
                 _pivotIndex = Utils.CreateArray(_inputSchema.Count, -1);
                 for (int i = 0; i < _infos.Length; i++)
                 {
                     var info = _infos[i];
-                    _pivotColMap[info.Name] = info.Index;
                     _ectx.Assert(_pivotIndex[info.Index] == -1);
                     _pivotIndex[info.Index] = i;
                 }
 
-                AsSchema = Schema.Create(this);
+                var schemaBuilder = new SchemaBuilder();
+                // Iterate through input columns. Input columns which are not pivot columns will be copied to output schema with the same column index unchanged.
+                // Input columns which are pivot columns would also be copied but with different data types and different metadata.
+                for (int i = 0; i < _inputSchema.Count; ++i)
+                {
+                    if (_pivotIndex[i] < 0)
+                    {
+                        // i-th input column is not a pivot column. Let's do a naive copy.
+                        schemaBuilder.AddColumn(inputSchema[i].Name, inputSchema[i].Type, inputSchema[i].Metadata);
+                    }
+                    else
+                    {
+                        // i-th input column is a pivot column. Let's calculate proper type and metadata for it.
+                        var metadataBuilder = new MetadataBuilder();
+                        metadataBuilder.Add(inputSchema[i].Metadata, metadataName => ShouldPreserveMetadata(metadataName));
+                        // To explain the output type of pivot columns, let's consider a row
+                        //   Age UserID
+                        //   18  {"Amy", "Willy"}
+                        // where "Age" and "UserID" are column names and 20/{"Amy", "Willy"} is "Age"/"UserID" column in this example row.
```

Comment:
>0 [](start = 74, length = 1)

18, or is 20 the column index?

Reply:
It's 18. Thanks.

---
In reply to: [246127382](https://github.com/dotnet/machinelearning/pull/2066#discussion_r246127382) [](ancestors = 246127382)